curated_transformers/models/loading_util.py

```diff
@@ -0,0 +1,105 @@
+import itertools
+from typing import Mapping, Optional
+import torch
+from torch.nn import Module, Parameter
+
+
+def _emplace_module_tensor(
+    module: Module,
+    tensor_name: str,
+    tensor: torch.Tensor,
+    device: Optional[torch.device] = None,
+):
+    """Replaces a module's parameter or (persistent) buffer with the passed tensor and moves it
+    to the given device. This is a zero-copy operation (excluding D2H/H2D transfers) where the
+    input tensor is directly associated with the module. Unexpected behaviour can occur if the same
+    tensor is associated with multiple modules.
+
+    :param module: Module whose associated tensor needs replacing.
+    :param tensor_name: Name of the tensor to be replaced.
+    :param tensor: eplacement tensor.
+    :param device: Device to which the replacement tensor is moved.
+    """
+    is_parameter = tensor_name in module._parameters
+    is_buffer = tensor_name in module._buffers
+    assert is_parameter ^ is_buffer
+
+    if device is not None:
+        tensor = tensor.to(device=device)
+
+    size_mismatch_msg = (
+        "Expected the size of the replacement for `{}` to be {}, but got {}"
+    )
+
+    if is_parameter:
+        old_param = module._parameters[tensor_name]
+        assert old_param is not None
+        if old_param.shape != tensor.shape:
+            raise ValueError(
```

Comment:
Do we also need to check that the dtype matches?

Reply:
Good question. It's probably a good idea, but once implement quantized weight loading, the types won't match anyway. So, I think we can dispense with it?

Reply:
Added the check.